src/linker/Linker/Driver.cs

```diff
@@ -684,33 +676,41 @@ protected int SetupContext (ILogger customLogger = null)
 			return 0;
 		}
 
-		public bool Run (ILogger customLogger = null)
+		// Returns the exit code of the process. 0 indicates success.
+		// Known non-recoverable errors (LinkerFatalErrorException) set the exit code
+		// to the error code.
+		// May propagate exceptions, which will result in the process getting an
+		// exit code determined by dotnet.
+		public int Run (ILogger customLogger = null)
 		{
 			int setupStatus = SetupContext (customLogger);
 			if (setupStatus > 0)
-				return true;
+				return 0;
 			if (setupStatus < 0)
-				return false;
+				return 1;
 
 			Pipeline p = context.Pipeline;
 			PreProcessPipeline (p);
 
 			try {
 				p.Process (context);
-			} catch (Exception ex) {
-				if (ex is LinkerFatalErrorException lex) {
-					context.LogMessage (lex.MessageContainer);
-					Console.Error.WriteLine (ex.ToString ());
-				} else {
-					context.LogError ($"IL Linker has encountered an unexpected error. Please report the issue at https://github.com/mono/linker/issues \n{ex}", 1012);
-				}
-
-				return false;
+			} catch (LinkerFatalErrorException lex) {
+				context.LogMessage (lex.MessageContainer);
+				Console.Error.WriteLine (lex.ToString ());
+				Debug.Assert (lex.MessageContainer.Category == MessageCategory.Error);
+				Debug.Assert (lex.MessageContainer.Code != null);
+				return lex.MessageContainer.Code ?? 1;
```

Comment:
Do we have a check that Code is never set to 0 ?

Reply:
The `MessageContainer` ctors that set Code ensure it's in a valid range.